src/include/storage/write_ahead_log/log_manager.h

```diff
@@ -52,41 +105,267 @@ class LogManager {
   }
 
   /**
-   * Process all the accumulated log records and serialize them out to disk. A flush will always happen at the end.
-   * (Beware the performance consequences of calling flush too frequently) This method should only be called from a
-   * dedicated
-   * logging thread.
+   * For testing only
+   * @return number of buffers used for logging
    */
-  void Process();
+  uint64_t TestGetNumBuffers() { return num_buffers_; }
 
   /**
-   * Flush the logs to make sure all serialized records before this invocation are persistent. Callbacks from committed
-   * transactions are also invoked when possible. This method should only be called from a dedicated logging thread.
+   * Set the number of buffers used for buffering logs.
+   *
+   * The operation fails if the LogManager has already allocated more buffers than the new size
    *
-   * Usually this method is called from Process(), but can also be called by itself if need be.
+   * @param new_num_buffers the new number of buffers the log manager can use
+   * @return true if new_num_buffers is successfully set and false the operation fails
    */
-  void Flush();
+  bool SetNumBuffers(uint64_t new_num_buffers) {
+    if (new_num_buffers >= num_buffers_) {
+      // Add in new buffers
+      for (size_t i = 0; i < new_num_buffers - num_buffers_; i++) {
+        buffers_.emplace_back(BufferedLogWriter(log_file_path_));
+        empty_buffer_queue_.Enqueue(&buffers_[num_buffers_ + i]);
+      }
+      num_buffers_ = new_num_buffers;
+      return true;
+    }
+    return false;
+  }
 
  private:
+  friend class DiskLogConsumerTask;
+  friend class LogSerializerTask;
+  friend class LogFlusherTask;
+
+  // Flag to tell us when the log manager is running or during termination
+  bool run_log_manager_;
+
+  // System path for log file
+  const char *log_file_path_;
+
+  // Number of buffers to use for buffering logs
+  uint64_t num_buffers_;
+
   // TODO(Tianyu): This can be changed later to be include things that are not necessarily backed by a disk
-  // (e.g. logs can be streamed out to the network for remote replication)
-  BufferedLogWriter out_;
+  //  (e.g. logs can be streamed out to the network for remote replication)
   RecordBufferSegmentPool *buffer_pool_;
 
   // TODO(Tianyu): Might not be necessary, since commit on txn manager is already protected with a latch
   common::SpinLatch flush_queue_latch_;
   // TODO(Tianyu): benchmark for if these should be concurrent data structures, and if we should apply the same
-  // optimization we applied to the GC queue.
+  //  optimization we applied to the GC queue.
   std::queue<RecordBufferSegment *> flush_queue_;
 
   // These do not need to be thread safe since the only thread adding or removing from it is the flushing thread
   std::vector<std::pair<transaction::callback_fn, void *>> commits_in_buffer_;
 
+  // This stores all the buffers the serializer or the log consumer threads use
+  std::vector<BufferedLogWriter> buffers_;
```

Comment:
nit:  can you also break the logic into the two tasks? It should be cleaner that way.

Reply:
Done.